### PR TITLE
support 3D rendering shapes layer

### PR DESCRIPTION
**Pull request by arokem**
# Description
This is done by injecting some of the triangulation code from the
vectors module into shapes.

At this point, things are breaking because zooms and thumbnails are
assumed to be 2D in shapes.
## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?
See changes in the examples.

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
Thanks @arokem i'll take a look at this

**Comment by sofroniewn** [comment]
@arokem I've now gotten the basics working with the api we talked about. To see an example look at `examples/3D_paths.py`. I still need to add some unit tests, raise some not implemented errors if someone tries to do 3D rendering with other Shape types that are not just paths etc.

We should also think if we want to try and make the path triangulation a little nicer, but at this point that's a fairly well isolated piece of coding that needs to be done.

Here's a gif of the current performance:
3D_paths

Maybe if you could give this a go with some real brain data (and ideally send me over some of the brain data, including both the array for the grayscale images and the arrays for the paths) that would be great.

To bring @jni and others up to speed here, @arokem wanted to extending the `Path` shape type from the `Shapes` layer to render in 3D so he could overlay some paths on 3D images

**Comment by sofroniewn** [comment]
After a little bit of digging around https://github.com/vispy/vispy/blob/master/vispy/visuals/tube.py I found some very nice code for 3D tube rendering that was easy to port over so our 3D lines now look quite nice. I should probably port this over to the vectors layer too so our 3D vectors look nicer as well:

3D_paths_nice

**Comment by sofroniewn** [comment]
I've now extended this so something vaguely sensible will happen if you switch to 3D rendering regardless of what shapes are added (apart from paths it will render the shape in 3D but confined to the plane it was drawn in). We can support move advanced shapes like actual spheres or cubes later, but this should at least prevent everything from crashing.

Also it seems like vispy is having some "z_order" issues with respect to the shapes in this gif. Some of that may have been fixed by changes I've made in the add_surface PR #503, and I'd like to get that merged before trying to address any of the ordering problems here:

3D_polys

**Comment by sofroniewn** [comment]
I've now fixed a z-ordering bug that was present in the shapes layer too (and causing problems in the gif above). I've also fixed it on the points layer.

I've just got add tests and then this will be ready for review.

**Comment by arokem** [comment]
Looks very nice. I would like to try with some brain streamlines. 

But: I am not managing to rotate this in 3D -- only scroll through the slices. Am I supposed to be holding down a particular modifier key? Or click on something to be able to rotate the volume in 3D?

**Comment by sofroniewn** [comment]
`command-Y` will enter 3D rendering mode. we'll add a button to the UI soon

**Comment by sofroniewn** [comment]
I have now added comprehensive tests to this PR. It is ready for review. One thing I want to call out is right now we depend on a `_` function from vispy - `from vispy.visuals.tube import _frenet_frames` for generating the tubes. We could copy this over to our shape_utils or just leave it.

@arokem if you can send me so data in numpy array / list of numpy array like format then i can try giving it a vizualization and adding it to our gallery. Also if there are other features that might be of use to you let us know.

**Comment by arokem** [comment]
Neato! The secret keyboard combo! 

FYI: On a mac, ctrl-y does not switch over to 3D viewing, but command-y does.

**Comment by arokem** [comment]
@sofroniewn : I added some code that will pull down data in https://github.com/arokem/napari/pull/1. You'll need to `pip install dipy` (or similar) for this to work.

File: examples/3D_paths.py
Comment (jni) [inline examples/3D_paths.py]: Make up your mind! =P

File: napari/_vispy/vispy_shapes_layer.py
Comment (jni) [inline napari/_vispy/vispy_shapes_layer.py:48]: As usual: can you add a comment that the `::-1` converts numpy-style coordinates to xyz?

File: napari/layers/shapes/shape_list.py
Comment (jni) [inline napari/layers/shapes/shape_list.py:385]: Currently, all the shapes appear on top of the MIP of the image/volume layer. Is this our fault or a limitation of vispy? I would expect/want the shapes to be in among the binary blobs.

**Review by jni — CHANGES_REQUESTED** [review]
@arokem @sofroniewn this is so cool! The 3D rendering feels like a work of art! :grinning: 

However, currently there appears to be a big bug: in `examples/nD_shapes.py` if I don't delete the broadcasted shapes layer before switching to 3D viewing, I get a very nasty traceback in the console, which includes this segment twice:

```pytb
WARNING: Traceback (most recent call last):
  File "examples/nD_shapes.py", line 52, in <module>
    f'broadcast: nshapes {layer.nshapes}, mask shape {masks.shape}, '
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/home/jni/projects/napari/napari/_qt/event_loop.py", line 21, in gui_qt
    app.exec_()
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 501, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 495, in keyPressEvent
    self._keyEvent(self._vispy_canvas.events.key_press, ev)
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 544, in _keyEvent
    func(native=ev, key=key, text=text_type(ev.text()), modifiers=mod)
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "/home/jni/projects/napari/napari/_qt/qt_viewer.py", line 337, in on_key_press
    gen = func(parent)
  File "/home/jni/projects/napari/napari/keybindings.py", line 38, in toggle_ndisplay
    viewer.dims.ndisplay = 3
  File "/home/jni/projects/napari/napari/components/dims.py", line 216, in ndisplay
    self.events.ndisplay()
  File "/home/jni/projects/napari/napari/util/event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "/home/jni/projects/napari/napari/util/event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "/home/jni/projects/napari/napari/util/event.py", line 523, in _invoke_callback
    cb(event)
  File "/home/jni/projects/napari/napari/components/viewer_model.py", line 83, in <lambda>
    self.dims.events.ndisplay.connect(lambda e: self._update_layers())
  File "/home/jni/projects/napari/napari/components/viewer_model.py", line 1013, in _update_layers
    layer.dims.ndisplay = self.dims.ndisplay
  File "/home/jni/projects/napari/napari/components/dims.py", line 216, in ndisplay
    self.events.ndisplay()
  File "/home/jni/projects/napari/napari/util/event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "/home/jni/projects/napari/napari/util/event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "/home/jni/projects/napari/napari/util/event.py", line 523, in _invoke_callback
    cb(event)
  File "/home/jni/projects/napari/napari/layers/base/base.py", line 167, in <lambda>
    self.dims.events.ndisplay.connect(lambda e: self._update_dims())
  File "/home/jni/projects/napari/napari/layers/base/base.py", line 315, in _update_dims
    self._set_view_slice()
  File "/home/jni/projects/napari/napari/layers/shapes/shapes.py", line 673, in _set_view_slice
    self._data_view.ndisplay = self.dims.ndisplay
  File "/home/jni/projects/napari/napari/layers/shapes/shape_list.py", line 103, in ndisplay
    self.add(shape, shape_index=index)
  File "/home/jni/projects/napari/napari/layers/shapes/shape_list.py", line 212, in add
    self._vertices, shape.data_displayed, axis=0
  File "<__array_function__ internals>", line 6, in append
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/numpy/lib/function_base.py", line 4700, in append
    return concatenate((arr, values), axis=axis)
  File "<__array_function__ internals>", line 6, in concatenate
ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 3 and the array at index 1 has size 2
```

and this segment repeated many times:

```pytb
WARNING: Traceback (most recent call last):
  File "examples/nD_shapes.py", line 52, in <module>
    f'broadcast: nshapes {layer.nshapes}, mask shape {masks.shape}, '
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/home/jni/projects/napari/napari/_qt/event_loop.py", line 21, in gui_qt
    app.exec_()
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 501, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 469, in mouseMoveEvent
    modifiers=self._modifiers(ev),
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/app/base.py", line 213, in _vispy_mouse_move
    ev = self._vispy_canvas.events.mouse_move(**kwargs)
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/home/jni/miniconda3/envs/f7/lib/python3.7/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "/home/jni/projects/napari/napari/_qt/qt_viewer.py", line 298, in on_mouse_move
    self.layer_to_visual[layer].on_mouse_move(event)
  File "/home/jni/projects/napari/napari/_vispy/vispy_base_layer.py", line 178, in on_mouse_move
    self.layer.position = self._transform_position(self._position)
  File "/home/jni/projects/napari/napari/layers/base/base.py", line 282, in position
    self._update_coordinates()
  File "/home/jni/projects/napari/napari/layers/base/base.py", line 473, in _update_coordinates
    self._value = self.get_value()
  File "/home/jni/projects/napari/napari/layers/shapes/shapes.py", line 1098, in get_value
    shape = self._data_view.inside(coord)
  File "/home/jni/projects/napari/napari/layers/shapes/shape_list.py", line 721, in inside
    indices = inside_triangles(triangles - coord)
ValueError: operands could not be broadcast together with shapes (0,3,3) (2,)
```

If I delete the broadcast layer *before* switching to 3D, then everything works great. If I try to delete it after I've switched to 3D (where I get a 3D rendering but pan-zoom instead of an arcball), then napari crashes.

I think that needs to get fixed before merging *and* a test added to make sure the behaviour doesn't come back.

File: examples/3D_paths.py
Comment (sofroniewn) [reply examples/3D_paths.py]: fixed ;-)

File: napari/_vispy/vispy_shapes_layer.py
Comment (sofroniewn) [reply napari/_vispy/vispy_shapes_layer.py:48]: added

File: napari/layers/shapes/shape_list.py
Comment (sofroniewn) [reply napari/layers/shapes/shape_list.py:385]: that's sort of our fault right now because of layer's concept would be hard to change - and definitely not something for this PR.

**Comment by sofroniewn** [comment]
@jni i've made the minor requests that you made. the issue with 2D layers viewed in 3D needs to be solved across all our layers and I'd rather do that in a follow up PR than this one. If you agree can you approve and merge.

**Comment by sofroniewn** [comment]
This PR now fixes #547 - you can transition to 3D rendering of all our layers including 2D ones, they just appear as very thin slices through z=0. The only thing that won't render in 3D right now is an rgb image. We could add special handling to convert it to grayscale and render it (this would be very easy!) or we could think about rendering it as three separate channels with red green and blue, but this would much more annoying. Thoughts @jni? This is now ready for re-review

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
@sofroniewn I've approved so you can get on with your work, but your fix for #547 probably needs a test, either in this PR or a follow-up one (in which case, make sure 547 remains open after merging this).

**Comment by sofroniewn** [comment]
I have now added comprehensive tests that will catch the bugs in #547. I will merge once the tests pass

**Pull request merged**
